src/Mapper.cpp

```diff
@@ -1346,3 +1346,223 @@ map<int, CGRANode*>* Mapper::getReorderPath(map<CGRANode*, int>* t_path) {
+                return -1;
+        }
+        json refs;
+        refFile >> refs;
+        CGRANodeID2Level.clear();
+        for(list<DFGNode*>::iterator dfgNode=t_dfg->nodes.begin(); dfgNode != t_dfg->nodes.end(); ++dfgNode){
+                int dfgNodeID = (*dfgNode)->getID();
+                int x = refs["Opt2TileXY"][to_string(dfgNodeID)]["x"];
+                int y = refs["Opt2TileXY"][to_string(dfgNodeID)]["y"];
+                refMapRes[*dfgNode] = t_cgra->nodes[y][x];
+                int cgraNodeID = t_cgra->nodes[y][x]->getID();
+                CGRANodeID2Level[cgraNodeID] = refs["Tile2Level"][to_string(cgraNodeID)];
+        }
+
+        /*cout<<"Printing references read from increMapInput.json"<<endl;
+        for(map<DFGNode*, CGRANode*>::iterator iter = refMapRes.begin(); iter != refMapRes.end(); ++iter){
+                cout<<iter->first->getOpcodeName()<<iter->first->getID()<<"->Tile"<<iter->second->getID()<<", level="<<CGRANodeID2Level[iter->second->getID()]<<endl;
+        }
+        cout<<"Done..."<<endl;*/
+
+        return 0;
+}
+
+// generates variables for incremental mapping
+void Mapper::sortAllocTilesByLevel(CGRA* t_cgra){
+        map<int, vector<CGRANode*>> FanIO_CGRANodes;
+        vector<int> FanIOs;
+        int numTiles = 0;
+        for (int i=0; i<t_cgra->getRows(); ++i) {
+                for (int j=0; j<t_cgra->getColumns(); ++j) {
+
+                        // record the number of FanIO for each tile. 
+                        CGRANode* currentCGRANode = t_cgra->nodes[i][j];
+                        if(currentCGRANode->isDisabled()) continue; // only record tiles that have DFGNodes mapped
+                        int FanIO = max(currentCGRANode->getInLinks()->size(), currentCGRANode->getOutLinks()->size());
+                        FanIO_CGRANodes[FanIO].push_back(currentCGRANode);
+
+                        // record FanIO in the list if it appears for the first time
+                        if (find(FanIOs.begin(), FanIOs.end(), FanIO) == FanIOs.end()){
+                                FanIOs.push_back(FanIO);
+                        }
+
+                        numTiles++;
+                }
+        }
+        std::sort(FanIOs.rbegin(), FanIOs.rend()); // sort FanIOs from big to small
+
+        //cout<<"Generating vairable CGRANodes_sortedByLevel"<<endl;
+        CGRANodes_sortedByLevel.clear();
+        for(int level = 0; level < FanIOs.size(); level++){
+                int FanIO = FanIOs[level];
+                vector<CGRANode*> tiles = FanIO_CGRANodes[FanIO];
+                // for(auto tile:tiles) cout<<"Level"<<level<<": Tile"<<tile->getID()<<endl;
+                CGRANodes_sortedByLevel.push_back(tiles);
+        }
+        cout<<"Done..."<<endl;
+}
+
+// generates the placement recommendation list for current DFGNode by referencing its placement in the former mapping results
+// Two principles: RPT & MBO
```

Comment:
Can you elaborate on these two principles?

Reply:
Reference Placement Tendency (RPT): if DFGNode0 is mapped on CGRANode5 in the referenced mapping results, we believe it should have the tendency to be mapped on the CGRANode that has the same or the closest FanIO with CGRANode5, so that we don't need to iteratively call calcaulateCost() like in heuristicMapper(), which saves lots of time and therefore accelerating the mapping.

Minimize Bypass Operations (MBO): Bypass operations occupy the CGRALink resources and may lead to the decrease of mapping quality. Therefore, on the basis of RPT, we further calculate the number of bypass operations for each CGRANode candidate required to communicate with its predecessors, and the CGRANode with fewer bypass operations has the higher priority to be mapped.

An example
<img width="481" alt="example" src="https://github.com/user-attachments/assets/655b56dc-e2ef-49de-a27a-68520abbaca0">
I will post the link of the whole paper after it's published

Reply:
Thanks, super clear!

Reply:
Plz just provide the full name for `RPT` and `MBO` for now.

Reply:
Has been corrected